File: dsa/week(3)/Copy_constructors.js
class Book {
    constructor(title, author) {
      this.Title = title;
      this.Author = author;
      this.RatesCounter = 2;
      this.Rates = new Array(this.RatesCounter);
      this.Rates[0] = 5;
      this.Rates[1] = 4;
    }
    destroy() {
      this.Rates = null;
    }
    display() {
      console.log(this.Author, this.RatesCounter);
      console.log(this.Rates ? this.Rates.length : 0);
    }
  
    // Shallow copy constructor
    shallowCopy(original) {
      this.Title = original.Title;
      this.Author = original.Author;
      this.RatesCounter = original.RatesCounter;
      this.Rates = original.Rates;
      this.Rates[0] = 5;
      this.Rates[1] = 4;
    }
  
    // Deep copy constructor
    deepCopy(original) {
      this.Title = original.Title;
      this.Author = original.Author;
      this.RatesCounter = original.RatesCounter;
      this.Rates = new Array(original.RatesCounter);
      this.Rates[0] = 5;
      this.Rates[1] = 4;
    }
  }
  
  // Main
  const book1 = new Book("Millionaire", "M.J. DeMarco");
  const book2 = new Book("Millionaire", "M.J. DeMarco");
  book2.display();
  
  const book3 = new Book("");
//   book3.shallowCopy(book2);
  book3.deepCopy(book2);
  book3.display();
  

File: dsa/week(3)/Copy_constructors.py
class Book:
    def __init__(self, title, author):
        self.Title = title
        self.Author = author
        self.RatesCounter = 2
        self.Rates = [0] * self.RatesCounter
        self.Rates[0] = 5
        self.Rates[1] = 4

    # Destructor
    def __del__(self):
        self.Rates = None

    def display(self):
        print(self.Author, self.RatesCounter)
        # Python equivalent to sizeof for length
        print(len(self.Rates) if self.Rates else 0)

    # Shallow copy constructor
    def shallow_copy(self, original):
        self.Title = original.Title
        self.Author = original.Author
        self.RatesCounter = original.RatesCounter
        self.Rates = original.Rates
        self.Rates[0] = 5
        self.Rates[1] = 4

    # Deep copy constructor
    def deep_copy(self, original):
        self.Title = original.Title
        self.Author = original.Author
        self.RatesCounter = original.RatesCounter
        self.Rates = [0] * original.RatesCounter
        self.Rates[0] = 5
        self.Rates[1] = 4


# Main
book1 = Book("Millionaire", "M.J. DeMarco")
book2 = Book("Millionaire", "M.J. DeMarco")
book2.display()

book3 = Book("", "")
book3.deep_copy(book2)
book3.display()
